robots/purePursuitController.h

```diff
@@ -0,0 +1,169 @@
+
+                std::vector<millimeter_t> p1,p2;
+                auto p1x = segmentStart.x() - x;
+                auto p1y = segmentStart.y() - y;
+                auto p2x = segmentEnd.x() - x;
+                auto p2y = segmentEnd.y() - y;
+
+                // intersection point with circle
+                auto dx = p2x - p1x;
+                auto dy = p2y - p1y;
+                auto d = sqrt(dx * dx + dy * dy);
+                auto D = p1x * p2y - p2x * p1y;
+
+                // if the discriminant is zero -> no intersection
+                auto discriminant = r * r * d * d - D * D;
+                if (discriminant.value() < 0) continue;
+
+                // x components of the intersection point
+                auto x1 = (D * dy + sgn(dy.value()) * dx * sqrt(discriminant)) / (d*d);
+                auto x2 = (D * dy - sgn(dy.value()) * dx * sqrt(discriminant)) / (d*d);
+                // y components of the intersection point
+                auto y1 = (-D * dx + fabs(dy) * sqrt(discriminant)) / (d*d);
+                auto y2 = (-D * dx - fabs(dy) * sqrt(discriminant)) / (d*d);
+
+                // if there is 2 intersections possible with the circle, we select the second one, 
+                // as that will be closer to the end point of the segment
+                bool validIntersection1 = (min(p1x, p2x) < x1 && x1 < max(p1x, p2x)) ||
+                                          (min(p1y, p2y) < y1 && y1 < max(p1y,p2y));
+                bool validIntersection2 = (min(p1x, p2x) < x2 && x2 < max(p1x, p2x)) ||
+                                          (min(p1y, p2y) < y2 && y2 < max(p1y,p2y));
+
+                // we always want the latest path segment point so if we have a 
+                // valid point, we delete the previous point
+                if (validIntersection1 || validIntersection2) lookahead.clear(); 
+                if (validIntersection1) {
+                    lookahead.push_back((x1 + x));
+                    lookahead.push_back((y1 + y));
+                }
+
+                // if there is a valid 2. intersection point, we keep that and remove
+                // the first one
+                if (validIntersection2) {
+                    if (lookahead.empty() || fabs(x1 - p2x) > fabs(x2 - p2x) ||
+                        fabs(y1 - p2y) > fabs(y2 - p2y)) {
+                            
+                        lookahead.clear();
+                        lookahead.push_back((x2 + x)); 
+                        lookahead.push_back((y2 + y));
+                    }
+                }
+            }
+            return lookahead;
+        }
+        return lookahead;
+    }
+
+private:
+
+    millimeter_t lookAheadDistance;                 // the distance to look ahead
+    millimeter_t wheelBase;                         // length between wheel bases
+    std::vector<Vector2<millimeter_t>> wayPoints;   // list of waypoint coordinates
```

Comment:
I'd rename these members to ``m_LookaheadDistance``, ``m_WheelBase`` and ``m_WayPoints`` to be in keeping with the style we use elsewhere in BoB robotics. It is handy to know if a variable is a member just by looking at the name.

Reply:
:+1: